File: NCoreUtils.Logging.Abstractions/LogMessage.fs
namespace NCoreUtils.Logging

open Microsoft.Extensions.Logging
open System

[<AbstractClass>]
type LogMessage (category : string, logLevel : LogLevel, eventId : EventId, exn : exn) =
  member val Timestamp = DateTimeOffset.UtcNow
  member val Category  = category
  member val LogLevel  = logLevel
  member val EventId   = eventId
  member val Exception = exn
  abstract AsyncLog : sink:ISink -> Async<unit>
  abstract Enqueue : queue:ISinkQueue -> unit

type LogMessage<'a> (category, logLevel, eventId, exn, state : 'a, formatter : Func<'a, exn, string>) =
  inherit LogMessage (category, logLevel, eventId, exn)
  member val State     = state
  member val Formatter = formatter
  override this.AsyncLog sink =
    sink.AsyncLog (this.Timestamp, this.Category, this.LogLevel, this.EventId, this.State, this.Exception, this.Formatter)
  override this.Enqueue queue =
    queue.Enqueue (this.Timestamp, this.Category, this.LogLevel, this.EventId, this.State, this.Exception, this.Formatter)

File: NCoreUtils.Logging.Abstractions/LogMessage.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;

namespace NCoreUtils.Logging;

public abstract class LogMessage(string category, LogLevel logLevel, EventId eventId, Exception? exception) : IDisposable
{
    public DateTimeOffset Timestamp { get; private set; } = DateTimeOffset.UtcNow;

    public string Category { get; private set; } = category;

    public LogLevel LogLevel { get; private set; } = logLevel;

    public EventId EventId { get; private set; } = eventId;

    public Exception? Exception { get; private set; } = exception;

    protected LogMessage Update(string category, LogLevel logLevel, EventId eventId, Exception? exception)
    {
        Timestamp = DateTimeOffset.UtcNow;
        Category = category;
        LogLevel = logLevel;
        EventId = eventId;
        Exception = exception;
        return this;
    }

    public abstract ValueTask LogAsync(ISink sink, CancellationToken cancellationToken = default);

    public abstract void Enqueue(ISinkQueue queue);

    protected virtual void Dispose(bool disposing) { /* noop */ }

    public void Dispose()
    {
        GC.SuppressFinalize(this);
        Dispose(disposing: true);
    }
}

public class LogMessage<TState>(
    string category,
    LogLevel logLevel,
    EventId eventId,
    Exception? exception,
    TState state,
    Func<TState, Exception?, string> formatter)
    : LogMessage(category, logLevel, eventId, exception)
{
    public TState State { get; private set; } = state;

    public Func<TState, Exception?, string> Formatter { get; private set; } = formatter;

    protected LogMessage<TState> Update(
        string category,
        LogLevel logLevel,
        EventId eventId,
        Exception? exception,
        TState state,
        Func<TState, Exception?, string> formatter)
    {
        base.Update(category, logLevel, eventId, exception);
        State = state;
        Formatter = formatter;
        return this;
    }

    public override ValueTask LogAsync(ISink sink, CancellationToken cancellationToken = default)
        => sink.LogAsync(this, cancellationToken);

    public override void Enqueue(ISinkQueue queue)
        => queue.Enqueue(this);
}